### PR TITLE
People/davidgamba/rmd 22906 add workflow tasks targetted

### PR DESCRIPTION
**Pull request by DavidGamba**
@tonytvo please review.

### CODE REVIEW

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflow.groovy [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import groovy.transform.CompileStatic
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+@CompileStatic
+class VeracodeWorkflow {
+    private static Logger log = LoggerFactory.getLogger(VeracodeWorkflow.class);
+
+    static void appWorkflow(VeracodeAPI veracodeAPI,
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflow.groovy:37]: I can see we have lots of duplication in appWorkflow here and in sandboxWorkflow
and based on your comment, comments is one of a good place to extract the block of code to function like
```
getBuildStatusOnce
createBuildIfPreviousWorkIsDone
uploadBuildsAndStartPreScan
startScanWithPreScanResults
```

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        return veracodeSetup.sandboxFilesToUpload
+    }
+
+    Set<String> getModuleWhitelist() {
+        veracodeSetup = project.findProperty("veracodeSetup") as VeracodeSetup
+        return veracodeSetup.moduleWhitelist
+    }
+
+    void run() {
+        Integer maxTries = Integer.parseInt((maxUploadAttempts != null) ? maxUploadAttempts : '10')
```
Comment (tonytvo) [inline src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:65]: I would move this default logic to constructor and another thing I should note, for some reasons, I see lots of cases where we need something like getValueOrDefaultWith(), so I think you could do something like this in constructor:
maxUploadAttempts = Integer.parseInt(getValueOfDefaultWith(project.findProperty("maxUploadAttempts)))

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflow.groovy (hunk 1 above)
Comment (DavidGamba) [reply src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflow.groovy:37]: I understand, but at the end of the day there is a different method call for sandbox tasks and app tasks so if I do those methods I end up adding an if statement in each method.
I decided to just do each appart.

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy (hunk 2 above)
Comment (tonytvo) [reply src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:65]: what's the different between these 3 variables:
- "${project.buildDir}/veracode"
- getFileSet()
- getModuleWhiteList()
it seems like they probably have something overlapped?

another thing is can we make object to encapsulate all of the parameters? or portion of the parameters? typically, we only want 3-5 parameters in the function parameters .... the # of parameters is a little bit too high

File: src/test/groovy/com/calgaryscientific/gradle/VeracodeWorkflowDOTest.groovy [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    File buildInfoFile = getResource('buildinfo-1.4.xml')
+    // Status after pre scan is completed
+    File buildInfoFilePreScanSuccess = getResource('buildinfo-1.4-preScanSuccess.xml')
+    // Status after scan completed
+    File buildInfoFileResultsReady = getResource('buildinfo-1.4-complete.xml')
+
+    File filelistFile = getResource('filelist-1.1.xml')
+    File preScanResultsFile = getResource('prescanresults-1.4.xml')
+
+    def 'Test appWorkflow when previous build has results ready'() {
```
Comment (tonytvo) [inline src/test/groovy/com/calgaryscientific/gradle/VeracodeWorkflowDOTest.groovy:45]: typically, the test names should some how reflect the expect behavior or examples you're trying to accomplish .... so if it is named something like given previous build results ready then work flow should start scan ....., this case I can see what you are trying to accomplish.....

File: src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy (hunk 2 above)
Comment (DavidGamba) [reply src/main/groovy/com/calgaryscientific/gradle/VeracodeWorkflowSandboxTask.groovy:65]: Is `getValueOfDefaultWith` a groovy built-in?

**Comment by DavidGamba** [comment]
@tonytvo lets address code duplication in another PR. If there are no other concerns I would like to move forward and get the build system on green again.

**Comment by tonytvo** [comment]
sure ... I'm glad we sit down and discussed yesterday, sounds like we had good conversation/rich discussion yesterday ....

**Review by tonytvo — APPROVED** [review]

**Pull request merged**